app/services/create_next_edition_service.rb

```diff
@@ -0,0 +1,54 @@
+class CreateNextEditionService < ApplicationService
+  def initialize(current_edition:, user:, discarded_edition: nil)
+    @current_edition = current_edition
+    @user = user
+    @discarded_edition = discarded_edition
+  end
+
+  def call
+    raise "Can only create a next edition from a live edition" unless current_edition.live
+
+    current_edition.update!(current: false)
+    EditDraftEditionService.call(next_edition, user, current: true, revision: next_revision)
+    AssignEditionStatusService.call(next_edition, user: user, state: :draft)
+    next_edition.save!
+    next_edition
+  end
+
+private
+
+  attr_reader :current_edition, :user, :discarded_edition
+
+  def next_edition
+    @next_edition ||= begin
+      document = current_edition.document
+
+      return discarded_edition if discarded_edition.present?
+
+      Edition.new(document: document,
+                  number: document.next_edition_number,
+                  created_by: user)
+    end
```

Comment:
This looks to have gone a bit wrong following its most recent change. If I recall correctly Pete's suggestion was to avoid any queries for document so doing the document assign before the return doesn't achieve that. 

It's also feels rather strange doing a return within a begin block. It looks like you're trying to escape the begin block whereas the return would terminate the method.

I'd suggest either doing:

```
def next_edition
  return discarded_edition if discarded_edition

  @next_edition ||= Edition.new(document: current_edition.document,
                                number: current_edition.document.next_edition_number,
                                created_by: user)
end
```

or switching it back to how it was before. Myself, I'm not concerned about the extra query because in the circumstances we expect this to be used [a document will already be cached](https://github.com/alphagov/content-publisher/blob/708dcd9552a63f6e884711032f9eb3b347958787/app/controllers/errors_controller.rb#L3)

Reply:
@kevindew Thanks - have revered to previous version.